mysql/resource_grant.go

```diff
@@ -650,11 +653,16 @@ func setDataFromGrant(grant MySQLGrant, d *schema.ResourceData) *schema.Resource
 		d.Set("tls_option", tableGrant.TLSOption)
 
 	} else if procedureGrant, ok := grant.(*ProcedurePrivilegeGrant); ok {
-		d.Set("database", fmt.Sprintf("%s %s.%s", procedureGrant.ObjectT, procedureGrant.Database, procedureGrant.CallableName))
-		d.Set("table", "*")
 		d.Set("grant", grant.GrantOption())
 		d.Set("tls_option", procedureGrant.TLSOption)
 
+		if _, ok := d.GetOk("database"); !ok {
```

Comment:
There are 2 ways to specify grants (showing it without quoting here):
1. database="PROCEDURE db", table="procedure_name"
2. database="PROCEDURE db.procedure_name", table="*"

I didn't notice this before, but how will we know, which one to set? How will we compare it with real values? We could have a custom diff function, but that's not here yet if I'm reading this correctly.

Furthermore, adding this condition is not really fixing anything. If for any reason database or table changes, we won't catch it.

Reply:
`setDataFromGrant` is used in two contexts:

1. We are importing a grant using `terraform import`
2. We are reading the remote state of an existing grant

In both contexts, the `database` and `table` should always be defined, so we don't actually need these at all. I added a note in the latest version of this PR.

Reply:
> I didn't notice this before, but how will we know, which one to set? How will we compare it with real values? We could have a custom diff function, but that's not here yet if I'm reading this correctly.

The overall behavior (right now) is:
- We preserve whatever value of `database` and `table` the user gives us in `ResourceData`.
- When we load from `ResourceData` (Terraform State), we construct a `MySqlGrant` interface by parsing `database` and `table`. If procedure is present in `database`, we specifically create a `ProcedurePrivilegeGrant` implementation of `MySqlGrant`, which knows how to parse both Example 1 and Example 2. 
- When we read from MySQL state, we also construct a  `MySqlGrant` interface by parsing `database` and `table` without modifying these properties.

Reply:
> Furthermore, adding this condition is not really fixing anything. 

This condition is fixing the test case that I added ([Test Case](https://github.com/petoju/terraform-provider-mysql/pull/104/files#diff-3458d1a380c2735ca9a16d3fbf7a34ccd4fe66a0025f3055ce7396bfbd09f055R862)). We basically were over-riding the existing value of `database` with Example 1's expected `database` semantics. There is no reason to do this.

> If for any reason database or table changes, we won't catch it.

If the user changes the name of `database` or `table` on the Terraform side, this is resolved correctly; both of these are labeled as `ForceNew` properties, and so Terraform will destroy and re-create the table accordingly.

If the user changes the name of `database` or `table` on the MySQL side, there is absolutely nothing we can do about it. We would need some capability to "detect" the new table name (a metastore or something), which is probably beyond the scope of this module; far easier to say, use Terraform.

Reply:
> If for any reason database or table changes, we won't catch it.

Ahh... but we do need to detect the case that the grant was deleted on the MySQL side, and that doesn't look properly handled either. Let me add a test + code.

Reply:
> If for any reason database or table changes, we won't catch it.

I think I see the issue you are talking about. We aren't actually reading in the state from MySQL. I am __shocked__ this isn't failing tests. I will make the needed changes.

Reply:
Also apologies for throwing this breakfix your way: I was relying on TDD, and am realizing these tests don't have great coverage.

Reply:
This looks like it was hiding a bunch of issues. It might be a good idea to label this as a broken version so people don't pick this up.

Reply:
> This looks like it was hiding a bunch of issues. It might be a good idea to label this as a broken version so people don't pick this up.

I cannot do that so easily AFAIK. I did rollback and released it with a new version so that people don't get the version we had. I'm fine with rolling that forward again - I am only buying us more time.

> I think I see the issue you are talking about. We properly delete grants if they are deleted (because all grants disappear), but we don't actually update the state if it is changed in the DB without Terraform. I am shocked this isn't failing tests. I will make the needed changes.

Hopefully it's something like this - but not only the grants are deleted, but some other do exist.
That said, as you started doubting, I started doubting about myself too, so I'm not 100% convinced this can happen given the code flow. Still, having a condition on setting attributes this may cause unpredictable behaviour.

> Also apologies for throwing this breakfix your way: I was relying on TDD, and am realizing these tests don't have great coverage.

No problem, I should have noticed. I also relied on TDD as I extended tests with everything where the previous provider failed, but it seems the coverage is not comprehensive.

The latest version is not setting those attributes at all - and that's even worse as imports will be broken

Reply:
@petoju Okay. I added some tests to cover refreshing; there ended up being a bunch more that needed fixing. 

What is the best course of action for rolling this forward?

Reply:
@DerekTBrown thanks for the tests!

I added some comments - they are not critical, just something to think about. Like for (not) setting `database` / `table` field - I prefer functions to just work independently from code paths that lead to them. It makes it much easier to reason about what the program does.

I'm a bit short on time, so this may take few days, but my plan now is to:
1. Please rebase this PR against latest main. Maybe you can copy over the files or so - there are no changes in your files.
2. I'll try this in my preexisting and a bit more complex environment. That may catch some issues esp. with diffs. My expectation is that this will cause no unexplained diff. I plan also trying some imports. That said, it's just another manual step and everything is probably well covered by tests.
3. If everything passes well, I'll merge this and release another version.

Reply:
> If everything passes well, I'll merge this and release another version.

Is there a mechanism to release it as an alpha version?

Reply:
Also: Could you enable squash before merge for this repo? I think it makes sense to squash PRs into a single commit.

Reply:
> Is there a mechanism to release it as an alpha version?

There is no explicit mechanism like that.
I mean I could post URLs here and ask you to download it. But once it's a real release, it's always of the same "stability" to users. I could release it as v 2.x, but that means rollback would still have to be 2.x.

I could release it as an older version - I haven't tried it and it would be almost certainly confusing. But that's it.

Reply:
> Also: Could you enable squash before merge for this repo?

Yes, I'll make it a proper merge or squash this time. I noticed that extra commits only when rolling it back.